esphome/components/sps30/sps30.cpp

```diff
@@ -0,0 +1,256 @@
+#include "sps30.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace sps30 {
+
+static const char *TAG = "sps30";
+
+static const uint16_t SPS30_CMD_GET_ARTICLE_CODE = 0xD025;
+static const uint16_t SPS30_CMD_GET_SERIAL_NUMBER = 0xD033;
+static const uint16_t SPS30_CMD_GET_FIRMWARE_VERSION = 0xD100;
+
+static const uint16_t SPS30_CMD_START_CONTINUOUS_MEASUREMENTS = 0x0010;
+static const uint16_t SPS30_CMD_START_CONTINUOUS_MEASUREMENTS_ARG = 0x0300;
+static const uint16_t SPS30_CMD_GET_DATA_READY_STATUS = 0x0202;
+static const uint16_t SPS30_CMD_READ_MEASUREMENT = 0x0300;
+static const uint16_t SPS30_CMD_STOP_MEASUREMENTS = 0x0104;
+static const uint16_t SPS30_CMD_SET_AUTOMATIC_CLEANING_INTERVAL_SECONDS = 0x8004;
+static const uint16_t SPS30_CMD_START_FAN_CLEANING = 0x5607;
+static const uint16_t SPS30_CMD_SOFT_RESET = 0xD304;
+static const size_t SERIAL_NUMBER_LENGTH = 8;
+
+void SPS30Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up sps30...");
+  this->write_command_(SPS30_CMD_SOFT_RESET);
+  delayMicroseconds(500000);
```

Comment:
```suggestion
  delay(500);
```

1. This is a very excessive delay, using `delay()` here will at least allow some background activity to occur (and the WDT won't go crazy).
2. Is this long timeout really needed? Does the device take that long to boot up?
3. If it is indeed half a second, then this needs to be handled differently - stalling the main thread for that long is not good. You can call SOFT_RESET from setup() and use `set_timeout` after that to delay the rest of the setup with an asynchronous timeout.

Reply:
Usually it does not, but in rare situations where the ESP32 restarts without repowering the sensor (i.g. when flashing new firmware) the initialization fails and the component gets marked as failed.
I will wrap the deferred logic into a an async block.